lib/auth/init.go

```diff
@@ -648,6 +660,94 @@ func createPresets(ctx context.Context, rm PresetRoleManager) error {
 	return nil
 }
 
+// PresetUserManager contains the required User Management methods to
+// create a preset User.
+type PresetUserManager interface {
+	// CreateUser creates a new user record based on the supplied `user` instance.
+	CreateUser(ctx context.Context, user types.User) error
+	// GetUser fetches a user from the repository by name, optionally fetching
+	// any associated secrets
+	GetUser(username string, withSecrets bool) (types.User, error)
+	// Upsert user creates or updates a user record as needed
+	UpsertUser(user types.User) error
+}
+
+func createPresetUsers(ctx context.Context, um PresetUserManager) error {
+	users := []types.User{
+		services.NewPresetAutomaticAccessBotUser(),
+	}
+	for _, user := range users {
+		// Some users are only valid for enterprise Teleport, and so will be
+		// nil for an OSS build and can be skipped
+		if user == nil {
+			continue
+		}
+
+		err := um.CreateUser(ctx, user)
+		if err == nil {
+			// Success! The rest of the loop body is immaterial. Move on to
+			// the next user.
+			continue
+		}
+
+		if !trace.IsAlreadyExists(err) {
+			return trace.Wrap(err, "failed creating preset user %s", user.GetName())
+		}
+
+		// Try to upgrade any existing user
+		oldUser, err := um.GetUser(user.GetName(), false)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		changed := migrateUser(oldUser, user)
+		if !changed {
+			// no changes, nothing to do here. Move along to the next preset user.
+			continue
+		}
+
+		if err = um.UpsertUser(oldUser); err != nil {
+			return trace.Wrap(err, "failed updating preset user %s", user.GetName())
+		}
+	}
+
+	return nil
+}
+
+// migrateUser attempts to update a user from an existing revision of
+// a preset user (aka `oldUser`) to match the latest revision of the
+// user preset (aka `newUser`) in place. Returns `true` if the `oldUser`
+// was modified in any way.
+//
+// Currently only updates the user's internal teleport labels (i.e. those
+// prefixed with TeleportInternalLabelPrefix).
+func migrateUser(oldUser, newUser types.User) bool {
```

Comment:
There's a discussion to be had whether any user with the system label should just be unilaterally replaced, and not go through any sort of migration that allows for user-side changes.

Reply:
Right, I would honestly just unconditionally upsert the user if the revision number is different to avoid complex migration/change detection logic.